src/timer.rs

```diff
@@ -0,0 +1,196 @@
+use crossbeam_channel::{Receiver, Sender};
+use std::{
+    thread::JoinHandle,
+    time::{Duration, Instant},
+};
+use uuid::Uuid;
+
+pub enum TimerType {
+    OneShot(Box<dyn FnOnce(ScheduleToken) + Send + 'static>),
+    Recurring(Duration, Box<dyn FnMut(ScheduleToken) -> TimerControlFlow + Send + 'static>),
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct ScheduleToken {
+    uuid: Uuid,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum TimerControlFlow {
+    Continue,
+    Cancel,
+}
+
+impl ScheduleToken {
+    fn new() -> Self {
+        let id = Uuid::new_v4();
+
+        Self { uuid: id }
+    }
+}
+
+enum ThreadMessage {
+    Shutdown,
+    Cancel(ScheduleToken),
+    Run(TimerEntry),
+}
+
+struct TimerEntry {
+    schedule_token: ScheduleToken,
+    scheduled_at: Instant,
+    timer_type: TimerType,
+}
+
+#[derive(Debug)]
+pub struct TimerHandle {
+    join_handle: JoinHandle<()>,
+    timer_ref: TimerRef,
+}
+
+#[derive(Debug, Clone)]
+pub struct TimerRef {
+    thread_tx: Sender<ThreadMessage>,
+}
+
+impl TimerRef {
+    pub fn run_once<F>(&mut self, delay: Duration, callback: F) -> ScheduleToken
```

Comment:
could this just call `run_recurring` internally with a wrapping closure that sends `TimerControlFlow::Cancel` or is that too convoluted? that way we wouldn't need `OneShot` and `Recurring`

Reply:
I'd prefer to keep it as-is, it feels nicer to use `FnOnce` for the "send once" case and `FnMut` just for the "send recurring" case.